feat(backup): Enable import over RPC

azaslavsky opened the pull request:
feat(backup): Enable imports over RPC

This finishes out the work started in #57740 and enables imports over
RPC as well. Like that PR, imports are already done on a sequential, per
model basis, so this change just consists of moving every such call
across an RPC boundary.

Closes getsentry/team-ospo#185
Closes getsentry/team-ospo#196
Closes getsentry/team-ospo#202

dashed commented on src/sentry/backup/helpers.py:
```diff
@@ -1,8 +1,11 @@
-from __future__ import annotations
+# Please do not use
+#     from __future__ import annotations
+# in modules such as this one where hybrid cloud data models or service classes are
```
This comment could be updated since this file isn't unique to hybrid cloud things.

azaslavsky replied:
Removed.

dashed commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,166 @@
+
+        try:
+            using = router.db_for_write(model)
+            with transaction.atomic(using=using):
```
@azaslavsky `ImportExportService` can only access tables that reside exclusively in the control silo (because of `local_mode = SiloMode.CONTROL`).

Do you anticipate that `import_export_service` have access to both the region and control silo resource?

If so, we'll want to define separate services for both the region and control silo.

An example can be seen here for `access_service`: https://github.com/getsentry/sentry/blob/c967fcead95f889efa5faff9c512c90613d0d66a/src/sentry/services/hybrid_cloud/access/service.py#L132-L138

azaslavsky replied:
Importing/exporting will always be run in the regional codepath when in siloed mode. So the only time we need to "go over the wire" will be when we are trying to import/export control models. Is there anything I could do to tell it to run the local impl vs using the remote one (I can do the impl-choosing logic myself, just not sure how to spell what I want).

Also, why did these tests not fail if I am calling this incorrectly?

dashed replied:
@azaslavsky Because the import tests are not marked with either `@region_silo_test(stable=True)` or `@region_silo_test(stable=True)` to establish the expected silo. Without these decorators, the test suite runs in monolith silo mode. Hence the tests won't fail.

I've added this snippet into `import_by_model()`, which caused some tests in `tests/sentry/backup/test_imports.py` to fail:

```python
        if in_test_environment():
            silo_limit = getattr(model._meta, "silo_limit", None)
            if not silo_limit:
                raise Exception("Expected silo_limit")
            if SiloMode.CONTROL not in silo_limit.modes:
                raise Exception("Accessed non-control model")
```

azaslavsky replied:
I've reworked this PR to work over RPC properly, with the `stable=True` annotations placed as needed.

azaslavsky committed:
feat(backup): Enable imports over RPC

This finishes out the work started in #57740 and enables imports over
RPC as well. Like that PR, imports are already done on a sequential, per
model basis, so this change just consists of moving every such call
across an RPC boundary.

Closes getsentry/team-ospo#185
Closes getsentry/team-ospo#196

azaslavsky commented:
@markstory This should be ready for review. The one failing test (colliding_org_auth_token) is fixed by https://github.com/getsentry/sentry/pull/58064, which I will land before this change.

markstory commented on src/sentry/services/hybrid_cloud/import_export/model.py:
```diff
@@ -0,0 +1,228 @@
+
+class RpcFilter(RpcModel):
+    """
+    Shadows `sentry.backup.helpers.Filter` for the purpose of passing it over an RPC boundary.
```
Would it be simpler to only have a filter object that can be serialized? Having to translate between two representations of the same concepts when neither is bound to the database could help simplify the modelling in backup/restore.

azaslavsky replied:
I already attempted this when originally writing this PR, but ran into some difficulty that I now wish I wrote down more clearly. I do remember time boxing that effort and then deciding that wrangling with it wasn't worth it, and leaving the TODO a few lines down on L37 to look at it later as part of the tidying up tasks for this project.

markstory commented on src/sentry/testutils/helpers/backups.py:
```diff
@@ -126,34 +127,57 @@ def reversed_dependencies():
+        connection = connections[using]
+
+        # For some reason, the tables for `SentryApp*` models don't get deleted properly here
+        # when using `model.objects.all().delete()`, so we have to call out to Postgres
```
This is because of the `ParanoidManager` these models use. The `ParanoidManager` overwrites the default queryset generation to include additional conditions. The models that use `ParanoidManager` also have a `with_deleted` manager instance that will query all records.

azaslavsky replied:
That makes sense! Fixed to select the correct manager instead.

azaslavsky commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,372 @@
+                            # For models that may have circular references to themselves (unlikely),
+                            # keep track of the new pk in the input map as well.
+                            nonlocal max_pk
+                            max_pk = item.pk
```
At the moment, yes (see: `queryset = model._base_manager.filter(q).order_by(pk_name)` in `yield_objects()`). However, I concede that this may not always be true (I've considered ordering by other globally unique fields like email or slug where possible), so I've added a check here to ensure we return the true max even if this changes.

azaslavsky commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,372 @@
+                    o = obj.object
+                    if o._meta.app_label not in EXCLUDED_APPS or o:
+                        if o.get_possible_relocation_scopes() & allowed_relocation_scopes:
+                            o = obj.object
```
Removed, and refactored the naming generally (see below).

azaslavsky commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,372 @@
+                counter = 0
+                for obj in deserialize("json", json_data, use_natural_keys=False):
+                    counter += 1
+                    o = obj.object
```
Renamed these to (hopefully) be a bit clearer.

azaslavsky commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,372 @@
+
+            # If we wrote at least one model, make sure to update the sequences too.
+            if counter > 0:
+                table = o._meta.db_table
+                seq = f"{table}_id_seq"
+                with connections[using].cursor() as cursor:
+                    cursor.execute(f"SELECT setval(%s, (SELECT MAX(id) FROM {table}))", [seq])
```
Is this a correctness or performance concern? I feel like we should be okay on correctness since this takes place in the same transaction as the model writes themselves (starts on L108). I could see this being a performance issue, but I (naively - I'm far from a postgres expert!) can't imagine that taking the MAX of an indexed column and updating the sequence counter will induce that much of a penalty relative to bulk writing models in a single transaction? The lock has to be acquired and held for some non-trivial amount of time for the whole import action anyway, so unless there's something uniquely penalizing about calling `setval`, my intuition is to run with it until we see concrete evidence of slowdowns. Wdyt?

azaslavsky committed:
Fixed markstory@ comments

azaslavsky commented on src/sentry/backup/exports.py:
```diff
@@ -86,92 +91,67 @@ def _export(
+    def get_exporter_for_model(model: Type[Model]):
+        if SiloMode.CONTROL in model._meta.silo_limit.modes:  # type: ignore
+            return import_export_service.export_by_model
+        return ImportExportService.get_local_implementation().export_by_model  # type: ignore
```
Done.

azaslavsky commented on src/sentry/services/hybrid_cloud/import_export/impl.py:
```diff
@@ -0,0 +1,372 @@
+        if SiloMode.CONTROL in model._meta.silo_limit.modes:
+            import_export_service.export_by_model(...)
+        else:
+            ImportExportService.get_local_implementation().export_by_model(...)
```
Done.

azaslavsky committed:
Fix database clearing logic

azaslavsky committed:
Update dependency fixtures

azaslavsky committed:
Remove redundantly inlined calls

markstory commented on src/sentry/utils/env.py:
```diff
@@ -1,5 +1,16 @@
+def is_split_db() -> bool:
+    if len(settings.DATABASES) != 1:  # type: ignore
+        return True
+    for db in settings.DATABASES.values():  # type: ignore
+        if db["NAME"] in {"region", "control"}:
+            return True
```
When the application runs in split silo mode, neither control or region connections exist. Having this available in a generic location like this could be a footgun for future us.

markstory commented on src/sentry/utils/env.py:
```diff
@@ -1,5 +1,16 @@
+
+
+def is_split_db() -> bool:
+    if len(settings.DATABASES) != 1:  # type: ignore
+        return True
```
In saas, this condition will alwasy be true as our production DB is spread across multiple hosts.

markstory reviewed (approved):
Had a few comments about potential traps in the split silo method, but otherwise this looks good to me.
